### PR TITLE
(PT-738) Add `policy` subcommands and `policies` config option

### PR DESCRIPTION
**Pull request by beechtom**
This adds the new `policy apply|new|show` subcommands and the `policies` config option.

### CODE REVIEW

File: lib/bolt/application.rb [hunk 1]
```diff
@@ -413,6 +413,170 @@ def list_plugins
...
+        raise Bolt::Error.new(
+          "The following policies cannot be loaded: '#{unloadable_policies.join("', '")}'. "\
+          "Policies must be a Puppet class saved to a project's or module's manifests directory.",
+          'bolt/unloadable-policy-error'
+        )
+      end
+
+      # Execute a single include statement with all the policies to apply them
+      # to the targets. Yay, reusable code!
+      # CODEREVIEW: Is there a meaningful difference between the one-line include
```
Comment (nicklewis) [inline lib/bolt/application.rb:467]: Nope! One line is perfect.

File: lib/bolt/project_manager.rb [hunk 2]
```diff
@@ -119,7 +119,7 @@ def create(path, name, modules)
         @outputter.stop_spin
       end
 
-      data = { 'name' => project_name }
+      data = { 'name' => project_name, 'profiles' => [] }
```
Comment (nicklewis) [inline lib/bolt/project_manager.rb:122]: Should be `policies`

Comment (beechtom) [reply lib/bolt/project_manager.rb]: Removed this altogether. Per a discussion in the design doc, we felt that while this feature is experimental we should not add it to projects by default.

Comment (nicklewis) [reply lib/bolt/project_manager.rb]: Oh yeah that makes way more sense 👍

**Review by nicklewis — APPROVED** [review]

File: lib/bolt/application.rb [hunk 3]
```diff
@@ -413,6 +413,170 @@ def list_plugins
...
+      unavailable_policies = policies.reject do |policy|
+        @config.policies.any? do |known_policy|
+          File.fnmatch?(known_policy, policy, File::FNM_EXTGLOB)
+        end
+      end
+
+      if unavailable_policies.any?
+        command = Bolt::Util.powershell? ? 'Get-BoltPolicy' : 'bolt policy show'
+
+        # CODEREVIEW: Phrasing
```
Comment (lucywyman) [inline lib/bolt/application.rb:436]: @hestonhoffman If you get a chance?

File: lib/bolt/application.rb [hunk 4]
```diff
@@ -413,6 +413,170 @@ def list_plugins
...
+      rescue Errno::EEXIST => e
+        raise Bolt::Error.new(
+          "#{e.message}; unable to create manifests directory '#{policy_dir}'",
+          'bolt/existing-file-error'
+        )
+      end
+
+      # Create the new policy
+      begin
+        # CODEREVIEW: Do we want a more detailed, commented out example of what a policy might look like?
```
Comment (lucywyman) [inline lib/bolt/application.rb:531]: I think this is fine for now, that might be something we add later.

File: lib/bolt/cli.rb [hunk 5]
```diff
@@ -287,6 +288,26 @@ def parse
         raise Bolt::CLIError, "Must specify a plan."
       end
 
+      if options[:subcommand] == 'policy'
+        if options[:action] == 'apply' && !options[:object]
```
Comment (lucywyman) [inline lib/bolt/cli.rb:292]: Need to add `subcommand policy` to line 348 below for `--noop` to work.

File: lib/bolt/bolt_option_parser.rb [hunk 6]
```diff
@@ -653,6 +669,61 @@ module            Manage Bolt project modules
...
+          policy apply
+
+      #{colorize(:cyan, 'Usage')}
+          bolt policy apply <policy> [options]
+
+      #{colorize(:cyan, 'Description')}
+          Apply a policy to the specified targets.
+    HELP
+
+    # CODEREVIEW: Do we want to have a more detailed description for the command
```
Comment (lucywyman) [inline lib/bolt/bolt_option_parser.rb:701]: Definitely not here. If we do anything it should be adding docs and linking to them here like we do for `bolt task` 
```
Documentation
    Learn more about Bolt tasks at http://pup.pt/bolt-tasks.
```

File: lib/bolt/cli.rb [hunk 7]
```diff
@@ -631,6 +652,26 @@ def execute(options)
         outputter.print_plugin_list(**app.list_plugins)
         SUCCESS
 
+      when 'policy'
```
Comment (lucywyman) [inline lib/bolt/cli.rb:655]: I mean so much hard work but this code block is beautiful 👨🏼‍🍳 💋

File: lib/bolt/application.rb [hunk 8]
```diff
@@ -413,6 +413,170 @@ def list_plugins
...
+      rescue Errno::EACCES => e
+        raise Bolt::FileError.new("#{e.message}; unable to create policy", policy)
+      end
+
+      # Update the project configuration to include the new policy
+      project_config = Bolt::Util.read_yaml_hash(@config.project.project_file, 'project config')
+
+      # Add the 'policies' key if it does not exist
+      project_config['policies'] ||= []
+      project_config['policies'] <<  name
```
Comment (lucywyman) [inline lib/bolt/application.rb:546]: There's a funny case here where if the *file* doesn't exist, but the policy is still listed here, it'll be included twice. I don't think it's a big deal, and definitely we shouldn't try to de-duplicate globs, but if it bugs you we can de-duplicate here too as well as when listing.

**Review by lucywyman — APPROVED** [review]
## Misc
- [x] JSON output works for all commands

## Help
- [x] `policy` shows up in `bolt help` list of subcommands
- [x] `bolt policy --help`
- [x] Warns when using the new subcommands
- [x] PowerShell cmdlets are generated
- [x] PowerShell cmdlets execute the underlying bolt commands

## `bolt policy show`
- [x] Prints what is in `bolt-project.yaml`
- [x] Errors if policies key is missing
- [x] Loads from the specified project path

## `bolt policy new`
- [x] Errors if the policy doesn't match the project name
- [x] Adds policy key if not present
- [x] Deduplicates the available policy list
- [x] Errors if the provided name is invalid
- [x] Supports subdirectories with `module::foo::bar` and creates the directory

## `bolt policy apply`
- [x] Errors if the policy specified doesn't exist
- [x] Errors if the policy specified isn't under `policies` key 
- [x] Loads policies from projects
- [x] Loads policies from the modulepath
- [x] Handles apply_prep failures
- [x] Handles apply failures
- [x] Supports globbing for allowing policies
- [x] Applies multiple policies in order
- [x] Uses `--noop`

I don't feel like this is exhaustive but I think it's close to good enough. Amazing work as always Tom!!

File: lib/bolt/application.rb (hunk 8 above)
Comment (beechtom) [reply lib/bolt/application.rb:543]: Updated to de-duplicate entries.

File: lib/bolt/application.rb [hunk 9]
```diff
@@ -413,6 +413,168 @@ def list_plugins
...
+        end
+      end
+
+      if unavailable_policies.any?
+        command = Bolt::Util.powershell? ? 'Get-BoltPolicy' : 'bolt policy show'
+
+        # CODEREVIEW: Phrasing
+        raise Bolt::Error.new(
+          "The following policies are not available to the project: '#{unavailable_policies.join("', '")}'. "\
+          "Policies must be listed in a project's 'policies' setting before they are applied to targets. "\
```
Comment (hestonhoffman) [inline lib/bolt/application.rb:439]:
```suggestion
          "You must list policies in a project's 'policies' setting before Bolt can apply them to targets. "\
```
Just taking out some passive voice

**Review by steveax — APPROVED** [review]

File: lib/bolt/application.rb [hunk 10]
```diff
@@ -413,6 +413,168 @@ def list_plugins
...
+    # @param targets [Array[String]] The list of targets to apply the policies to.
+    # @param noop [Boolean] Whether to apply the policies in no-operation mode.
+    # @return [Bolt::ResultSet]
+    #
+    def apply_policies(policies, targets, noop: false)
+      policies = policies.split(',')
+
+      # Validate that the policies are available to the project.
+      unavailable_policies = policies.reject do |policy|
+        @config.policies.any? do |known_policy|
```
Comment (lucywyman) [inline lib/bolt/application.rb:428]: I don't *think* people will do this, but if you run `be bolt policy apply project::file, -t docker` this stacktraces

File: lib/bolt/application.rb (hunk 3 above)
Comment (lucywyman) [reply lib/bolt/application.rb:436]: I think can probably remove the CODEREVIEWs?

File: lib/bolt/application.rb (hunk 10 above)
Comment (beechtom) [reply lib/bolt/application.rb]: Good catch. Updated.

**Review by lucywyman — APPROVED** [review]

**Pull request merged**